### PR TITLE
Add docstrings

### PR DESCRIPTION
**Pull request by AttackingOrDefending**
Not every parameter has a `:param parameter: parameter description` because I thought that some are self-explanatory. Some parameters are explained in some other parts of the code, so I didn't add a description to avoid repetition. If you think something else needs explaining I will add it. I'm not sure if all the descriptions are correct (mostly for the logging code at the start of `lichess-bot.py`, and the filters at the start of `config.py`).

Uses `flake8-docstrings` which uses PEP257 to find errors in the docstrings. It finds 158 D401 errors.
EDIT: Fixed the errors.

### CODE REVIEW

File: .github/workflows/python-build.yml [hunk 1]
```diff
@@ -28,14 +28,19 @@ jobs:
...
-        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide.
+        # W503 and W504 are mutually exclusive. W504 is considered the best practice now.
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics --ignore=D,W503
     - name: Lint with flake8-markdown
       run: |
         flake8-markdown "*.md"
+    - name: Lint with flake8-docstrings
+      run: |
+        # D212 and D213 conflict with other warni
```
Comment (MarkZH) [inline .github/workflows/python-build.yml:45]: This line is incomplete.

File: .github/workflows/python-build.yml [hunk 2]
```diff
@@ -28,14 +28,19 @@ jobs:
...
+        flake8 . --count --exit-zero --max-line-length=127 --statistics --select=D
```
Comment (MarkZH) [inline .github/workflows/python-build.yml:46]: This should probably ignore D212.

File: config.py [hunk 3]
```diff
@@ -12,41 +13,74 @@
 
 
 class FilterType(str, Enum):
+    """What to do if the opponent declines our challenge."""
+
     NONE = "none"
+    """Will still challenge the opponent."""
     COARSE = "coarse"
+    """Won't challenge the opponent for some time."""
```
Comment (MarkZH) [inline config.py:21]: Since bools have replaced timers, this should probably read: Won't challenge the opponent again.

File: config.py [hunk 4]
```diff
@@ -12,41 +13,74 @@
...
     FINE = "fine"
+    """Won't challenge the opponent to a game of the same mode, speed, and variant for some time."""
```
Comment (MarkZH) [inline config.py:23]: Replace "for some time" with "again based on the reason for the opponent declining the challenge."

File: .github/workflows/python-build.yml (hunk 1 above)
Comment (AttackingOrDefending) [reply .github/workflows/python-build.yml:45]: This line was left from when I was using `--docstring-convention all`, so the warning conflicted between the different standards. Now I use the default which is `--docstring-convention pep257`, so they don't conflict.

File: conversation.py [hunk 5]
```diff
@@ -21,11 +33,24 @@ def __init__(self, game: model.Game, engine: EngineWrapper, xhr: Lichess, versio
     command_prefix = "!"
 
     def react(self, line: ChatLine, game: model.Game) -> None:
+        """
+        Reacts to the messages sent.
```
Comment (MarkZH) [inline conversation.py:37]: "Reacts to the received messages." would be clearer.

File: lichess.py [hunk 6]
```diff
@@ -90,18 +114,42 @@ def api_get(self, endpoint_name: str, *template_args: str,
...
+        :return: lichess.org's response in a list of dicts.
+        """
         response = self.api_get(endpoint_name, *template_args, params=params)
         json_response: List[JSON_REPLY_TYPE] = response.json()
         return json_response
 
     def api_get_raw(self, endpoint_name: str, *template_args: str,
                     params: Optional[Dict[str, str]] = None, ) -> str:
+        """
+        Send a GET to lichess.org.
```
Comment (MarkZH) [inline lichess.py:146]: Send a GET to lichess.org that returns plain text (UTF-8).

File: lichess.py [hunk 7]
```diff
@@ -206,19 +301,23 @@ def get_ongoing_games(self) -> List[Dict[str, Any]]:
...
+        """Resign a game."""
         self.api_post("resign", game_id)
 
     def set_user_agent(self, username: str) -> None:
+        """Set the user agent for communication with lichess.org."""
         self.header.update({"User-Agent": f"lichess-bot/{self.version} user:{username}"})
         self.session.headers.update(self.header)
 
     def get_game_pgn(self, game_id: str) -> str:
+        """Get the pgn of a game."""
```
Comment (MarkZH) [inline lichess.py:313]: Get the PGN (Portable Game Notation) record of a game.

File: matchmaking.py [hunk 8]
```diff
@@ -100,25 +109,31 @@ def create_challenge(self, username: str, base_time: int, increment: int, days:
             return ""
 
     def update_daily_challenge_record(self) -> None:
-        # As the number of challenges in a day increase, the minimum wait time between challenges increases.
-        # 0   -  49 challenges --> 1 minute
-        # 50  -  99 challenges --> 2 minutes
-        # 100 - 149 challenges --> 3 minutes
-        # etc.
+        """
+        As the number of challenges in a day increase, the minimum wait time between challenges increases.
```
Comment (MarkZH) [inline matchmaking.py:113]: Add as first sentence: "Record timestamp of latest challenge and update minimum wait time."

File: model.py [hunk 9]
```diff
@@ -11,23 +12,28 @@
...
-        self.speed = c_info["speed"]
-        self.increment: int = c_info.get("timeControl", {}).get("increment")
-        self.base: int = c_info.get("timeControl", {}).get("limit")
-        self.days: int = c_info.get("timeControl", {}).get("daysPerTurn")
-        self.challenger = Player(c_info.get("challenger") or {})
-        self.opponent = Player(c_info.get("destUser") or {})
+    """Store information about a challenge."""
+
+    def __init__(self, challenge_info: Dict[str, Any], user_profile: Dict[str, Any]) -> None:
+        """:param user_profile: Information about our bot."""
```
Comment (MarkZH) [inline model.py:18]: Need docstring on parameter `challenge_info`. Other methods in the class are missing `param` docstrings.

File: model.py [hunk 10]
```diff
@@ -112,73 +134,93 @@ class Termination(str, Enum):
 
 
 class Game:
-    def __init__(self, json: Dict[str, Any], username: str, base_url: str, abort_time: int) -> None:
+    """Store information about a game."""
+
+    def __init__(self, game_info: Dict[str, Any], username: str, base_url: str, abort_time: int) -> None:
+        """:param abort_time: How long to wait before aborting the game."""
```
Comment (MarkZH) [inline model.py:140]: Lots of methods in the class are missing `param` docstrings.

File: strategies.py [hunk 11]
```diff
@@ -1,6 +1,4 @@
-"""
-Some example strategies for people who want to create a custom, homemade bot.
-"""
+"""Some example strategies for people who want to create a custom, homemade bot."""
```
Comment (MarkZH) [inline strategies.py:1]: Add sentence at end: "With these classes, bot makers will not have to implement the UCI or XBoard interfaces themselves."

File: engine_wrapper.py [hunk 12]
```diff
@@ -29,6 +30,12 @@
 
 @contextmanager
 def create_engine(engine_config: config.Configuration) -> Generator[EngineWrapper, None, None]:
+    """
+    Create the engine.
```
Comment (MarkZH) [inline engine_wrapper.py:34]: Add: Use in a with-block to automatically close the engine when exiting the game.

File: engine_wrapper.py [hunk 13]
```diff
@@ -343,61 +445,93 @@ def to_readable_key(stat: str) -> str:
...
         self.engine = chess.engine.SimpleEngine.popen_uci(commands, timeout=10., debug=False, setpgrp=False, stderr=stderr,
                                                           **popen_args)
         self.engine.configure(options)
 
     def stop(self) -> None:
+        """Tell the engine to stop searching."""
         self.engine.protocol.send_line("stop")
 
     def get_opponent_info(self, game: model.Game) -> None:
+        """Get the opponents info and sends it to the engine."""
```
Comment (MarkZH) [inline engine_wrapper.py:506]: opponent's

File: engine_wrapper.py [hunk 14]
```diff
@@ -426,11 +561,13 @@ def report_game_result(self, game: model.Game, board: chess.Board) -> None:
         self.engine.protocol.send_line(f"result {game.result()}{endgame_message}")
 
     def stop(self) -> None:
+        """Tell the engine to stop searching."""
         self.engine.protocol.send_line("?")
 
     def get_opponent_info(self, game: model.Game) -> None:
-        if (game.opponent.name and isinstance(self.engine.protocol, chess.engine.XBoardProtocol) and
-                self.engine.protocol.features.get("name", True)):
+        """Get the opponents info and sends it to the engine."""
```
Comment (MarkZH) [inline engine_wrapper.py:568]: opponent's

File: lichess-bot.py [hunk 15]
```diff
@@ -380,6 +445,19 @@ def start_game(event: EVENT_TYPE,
                correspondence_queue: CORRESPONDENCE_QUEUE_TYPE,
                active_games: Set[str],
                low_time_games: List[EVENT_GETATTR_GAME_TYPE]) -> None:
+    """
+    Start a game.
+
+    :param event: The gameStart event.
+    :param pool: The pool that the game is added to, so they can be run asynchronously.
```
Comment (MarkZH) [inline lichess-bot.py:452]: The thread pool that ...

File: lichess-bot.py [hunk 16]
```diff
@@ -614,6 +716,7 @@ def should_exit_game(board: chess.Board, game: model.Game, prior_game: Optional[
 
 def final_queue_entries(control_queue: CONTROL_QUEUE_TYPE, correspondence_queue: CORRESPONDENCE_QUEUE_TYPE,
                         game: model.Game, is_correspondence: bool) -> None:
+    """Log the game that ended or we disconnected from, and sends a `local_game_done` for the game."""
```
Comment (MarkZH) [inline lichess-bot.py:719]: Add: If this is an unfinished correspondence game, put it in a queue to resume later.

File: lichess-bot.py [hunk 17]
```diff
@@ -672,6 +777,15 @@ def tell_user_game_result(game: model.Game, board: chess.Board) -> None:
 
 def try_print_pgn_game_record(li: lichess.Lichess, config: Configuration, game: model.Game, board: chess.Board,
                               engine: engine_wrapper.EngineWrapper) -> None:
+    """
+    Call `print_pgn_game_record` to write the game to a PGN file and handles errors raised by it.
```
Comment (MarkZH) [inline lichess-bot.py:781]: "handles" --> "handle"

File: lichess-bot.py [hunk 18]
```diff
@@ -732,6 +855,12 @@ def print_pgn_game_record(li: lichess.Lichess, config: Configuration, game: mode
 
 
 def fill_missing_pgn_headers(game_record: chess.pgn.Game, game: model.Game) -> None:
+    """
+    Fill the headers missing by the lichess.org PGN with local headers from `game`.
```
Comment (MarkZH) [inline lichess-bot.py:859]: Fill in any missing headers in the PGN record provided by lichess.org with information from `game`.

**Comment by MarkZH** [comment]
I've finished reviewing. I'll merge once the last edits have been handled. Good work!

File: lichess-bot.py [hunk 19]
```diff
@@ -222,6 +275,7 @@ def lichess_bot_main(li: lichess.Lichess,
                       "logging_level": logging_level}
 
     recent_bot_challenges: DefaultDict[str, List[Timer]] = defaultdict(list)
+    one_game = True
```
Comment (MarkZH) [inline lichess-bot.py:278]: I don't think this should be here.

File: engine_wrapper.py [hunk 20]
```diff
@@ -343,61 +447,93 @@ def to_readable_key(stat: str) -> str:
...
         self.engine = chess.engine.SimpleEngine.popen_uci(commands, timeout=10., debug=False, setpgrp=False, stderr=stderr,
                                                           **popen_args)
         self.engine.configure(options)
 
     def stop(self) -> None:
+        """Tell the engine to stop searching."""
         self.engine.protocol.send_line("stop")
 
     def get_opponent_info(self, game: model.Game) -> None:
+        """Get the opponent's info and sends it to the engine."""
```
Comment (MarkZH) [inline engine_wrapper.py:508]: "sends" --> "send"

File: engine_wrapper.py [hunk 21]
```diff
@@ -426,11 +563,13 @@ def report_game_result(self, game: model.Game, board: chess.Board) -> None:
         self.engine.protocol.send_line(f"result {game.result()}{endgame_message}")
 
     def stop(self) -> None:
+        """Tell the engine to stop searching."""
         self.engine.protocol.send_line("?")
 
     def get_opponent_info(self, game: model.Game) -> None:
-        if (game.opponent.name and isinstance(self.engine.protocol, chess.engine.XBoardProtocol) and
-                self.engine.protocol.features.get("name", True)):
+        """Get the opponent's info and sends it to the engine."""
```
Comment (MarkZH) [inline engine_wrapper.py:570]: "sends" --> "send"

File: lichess-bot.py (hunk 19 above)
Comment (AttackingOrDefending) [reply lichess-bot.py]: Good catch! I forgot to remove it after testing.

**Review by MarkZH — APPROVED** [review]

**Pull request merged**